packages/toolpad-app/src/toolpad/AppEditor/PageTitleEditor.tsx

```diff
@@ -0,0 +1,66 @@
+import { TextField } from '@mui/material';
+import * as React from 'react';
+import { AppDom, PageNode } from '../../appDom';
+import { useDomApi } from '../AppState';
+import { update } from '../../utils/immutability';
+
+interface PageTitleEditorProps {
+  node: PageNode;
+}
+
+export default function PageTitleEditor({ node }: PageTitleEditorProps) {
+  const domApi = useDomApi();
+  const [pageTitleInput, setPageTitleInput] = React.useState(node.attributes.title);
+  const [isValid, setIsValid] = React.useState(true);
+
+  const handlePageTitleChange = React.useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => setPageTitleInput(event.target.value),
+    [],
+  );
+
+  const handleCommit = React.useCallback(() => {
+    if (!isValid) {
+      return;
+    }
+    domApi.update((appDom: AppDom) => {
+      return update(appDom, {
+        nodes: update(appDom.nodes, {
+          [node.id]: {
+            ...node,
+            attributes: { ...node.attributes, title: pageTitleInput },
+          },
+        }),
+      });
+    });
+  }, [domApi, node, pageTitleInput, isValid]);
+
+  const handleKeyPress = React.useCallback(
+    (event: React.KeyboardEvent<HTMLDivElement>) => {
+      if (event.code === 'Enter') {
+        handleCommit();
+      }
+    },
+    [handleCommit],
+  );
+
+  React.useEffect(() => {
```

Comment:
No need for creating new state and an effect here. You can just derive state as follows:

```tsx
const isValid = !!pageTitleInput
```

Personally I'd rather solve this by creating the error first as

```tsx
const inputError = React.useMemo(() => {
  if (!pageTitleInput) {
    return 'Input require'
  }
  if (pageTitleInput.length > MAX_TITLE_SIZE) {
    return 'Page title too long'
  }
  // ...
  return null
}, [pageTitleInput])

const isValid = !inputError
```

But I'm wondering, wouldn't this make it impossible to remove a title from a page?

Reply:
If we link `validateInput` with `handleCommit`. It should work.